### PR TITLE
Build File Generator: Add BuildFileGenerator

### PR DESCRIPTION
**Pull request by val-fed**

### CODE REVIEW

File: tools/build_file_generator/BuildFileGenerator.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    for (ProtoFile protoFile : protoFiles) {
+      if ((imprt.getPackage() + "." + imprt.getClassName())
+          .startsWith(protoFile.getJavaPackage() + "." + protoFile.getJavaOuterClassname())) {
+        return path + ":" + protoFile.getFileName() + "_java_proto";
+      }
+    }
+    return path + ":" + imprt.getClassName();
+  }
+
+  private List<ProtoFile> getWholeProjectProtoFiles() {
```
[the remark this answers is not in the document]
Comment (val-fed) [reply tools/build_file_generator/BuildFileGenerator.java]: Done.

File: tools/build_file_generator/BuildFileGenerator.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      }
+
+      String target;
+      if (thirdPartyTargets.isEmpty()) {
+        // it isn't third party dep
+        target = getInternalDep(imprt, getWholeProjectProtoFiles());
+      } else if (thirdPartyTargets.size() == 1) {
+        target = thirdPartyTargets.get(0).getTarget();
+      } else {
+        // we choose the smallest
```
[the remark this answers is not in the document]
Comment (val-fed) [reply tools/build_file_generator/BuildFileGenerator.java]: Done.

File: tools/build_file_generator/BuildFileGenerator.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  // TODO: Implement method
+  private List<String> getResources() {
+    return new ArrayList<>();
+  }
+
+  private List<String> getDeps(JavaClass javaClass) {
+    List<String> result = new ArrayList<>();
+
+    for (Import imprt : javaClass.getImportList()) {
+      if (imprt.getPackage().startsWith("java") || imprt.getPackage().startsWith("javax")) {
```
[the remark this answers is not in the document]
Comment (val-fed) [reply tools/build_file_generator/BuildFileGenerator.java]: Yes, you are right. Thank you!

File: tools/build_file_generator/BuildFileWriter.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import com.google.startupos.tools.buildfilegenerator.Protos.BuildFile.JavaBinary;
+import com.google.startupos.tools.buildfilegenerator.Protos.BuildFile.JavaLibrary;
+import com.google.startupos.tools.buildfilegenerator.Protos.BuildFile.JavaProtoLibrary;
+import com.google.startupos.tools.buildfilegenerator.Protos.BuildFile.JavaTest;
+import com.google.startupos.tools.buildfilegenerator.Protos.BuildFile.ProtoLibrary;
+
+import java.util.List;
+import javax.inject.Inject;
+
+public class BuildFileWriter {
```
[the remark this answers is not in the document]
Comment (val-fed) [reply tools/build_file_generator/BuildFileWriter.java]: Done.

File: tools/build_file_generator/BuildFileGenerator.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+  // TODO: Implement method
+  private List<String> getResources() {
+    return new ArrayList<>();
+  }
+
+  private List<String> getDeps(JavaClass javaClass) {
+    List<String> result = new ArrayList<>();
+
+    for (Import imprt : javaClass.getImportList()) {
```
[the remark this answers is not in the document]
Comment (val-fed) [reply tools/build_file_generator/BuildFileGenerator.java]: Done.

File: tools/build_file_generator/BuildFileGenerator.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+public class BuildFileGenerator {
+  private static final String PATH_TO_ZIP = "tools/build_file_generator/third_party_deps.zip";
```
[the remark this answers is not in the document]
Comment (val-fed) [reply tools/build_file_generator/BuildFileGenerator.java]: Done.

File: common/FileUtils.java [hunk 7]
```diff
@@ -447,5 +450,28 @@ public void writePrototxtToZip(Message proto, String zipFilePath, String pathIns
...
   }
+
+  /** Reads a prototxt file into a proto from zip archive. */
+  public Message readPrototxtFromZip(
+      String zipFilePath, String pathInsideZip, Message.Builder builder) throws IOException {
+    ZipFile zipFile =
+        new ZipFile(expandHomeDirectory(joinPaths(getCurrentWorkingDirectory(), zipFilePath)));
+    ZipEntry entry = zipFile.getEntry(pathInsideZip);
+    InputStream inputStream = zipFile.getInputStream(entry);
+    String content =
```
[the remark this answers is not in the document]
Comment (val-fed) [reply common/FileUtils.java]: Done.

File: tools/build_file_generator/BuildFileGenerator.java [hunk 8]
```diff
@@ -168,12 +167,12 @@ private JavaTest getJavaTest(JavaClass javaClass) {
...
       if (thirdPartyTargets.isEmpty()) {
-        // it isn't third party dep
-        target = getInternalDep(imprt, getWholeProjectProtoFiles());
+        // It isn't third party dep.
+        target = getInternalDep(importProto, getWholeProjectProtoFiles());
       } else if (thirdPartyTargets.size() == 1) {
         target = thirdPartyTargets.get(0).getTarget();
       } else {
-        // we choose the smallest
+        // If the class exists in several third party deps we choose the smallest one.
```
Comment (oferb) [inline tools/build_file_generator/BuildFileGenerator.java]: I'd add: "since the other ones probably contain it"

Comment (val-fed) [reply tools/build_file_generator/BuildFileGenerator.java]: Done.

**Pull request merged**